### PR TITLE
8278241: Implement JVM SpinPause on linux-aarch64

### PR DESCRIPTION
**Pull request by eastig**
This JVM SpinPause uses a spin wait stub. The stub is generated based on the `SpinWait` description which is defined with `OnSpinWaitInst`/`OnSpinWaitInstCount` options.  The `SpinWait` provides the description of the instruction and the instruction count.

The `SpinWait` description is also used for the `_onSpinWait()` intrinsic. We don't have use cases when we need different implementations for the `_onSpinWait()` intrinsic and JVM SpinPause.

Testing results for fastdebug and release builds:
- `gtest`: Passed
- `tier1`...`tier4`: Passed
- `hotspot/jtreg/runtime/Thread/TestSpinPause.java`: Passed

JVM SpinPause is used for the synchronised statements and can benchmarked with `org.openjdk.bench.vm.lang.LockUnlock.testContendedLock`.

Benchmarking results (number of samples per an experiment: 150) for Graviton2 (Neoverse N1), 1 ISB instruction:

```

+-----------+-------------------+------------+-----------+-----------+----------+---------+
| CPU cores | Contended threads | Base ns/op |   Error   |    New    |  Error   |  Diff   |
+-----------+-------------------+------------+-----------+-----------+----------+---------+
|         8 |                64 |  10007.213 | ±910.911  |  8527.346 | ±377.242 | -14.79% |
|        16 |                64 |  10274.935 | ±880.568  |  8310.433 | ±326.845 | -19.12% |
|        32 |                64 |  12231.947 | ±1525.364 |  9205.941 | ±394.409 | -24.74% |
|        64 |                64 |    9929.49 | ±586.074  | 10488.695 | ±570.458 | 5.63%   |
|        64 |                32 |   5605.119 | ±629.340  |  5023.882 | ±230.639 | -10.37% |
|        64 |                16 |   2817.346 | ±263.696  |  2367.528 | ±94.158  | -15.97% |
|        64 |                 2 |    870.389 | ±530.579  |   464.395 | ±126.260 | -46.65% |
+-----------+-------------------+------------+-----------+-----------+----------+---------+
```

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8278241](https://bugs.openjdk.java.net/browse/JDK-8278241): Implement JVM SpinPause on linux-aarch64

### Reviewers
 * [Andrew Haley](https://openjdk.java.net/census#aph) (@theRealAph - **Reviewer**)
 * [Paul Hohensee](https://openjdk.java.net/census#phh) (@phohensee - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/6803/head:pull/6803` \
`$ git checkout pull/6803`

Update a local copy of the PR: \
`$ git checkout pull/6803` \
`$ git pull https://git.openjdk.java.net/jdk pull/6803/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 6803`

View PR using the GUI difftool: \
`$ git pr show -t 6803`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/6803.diff">https://git.openjdk.java.net/jdk/pull/6803.diff</a>

</details>

### CODE REVIEW

**Comment by eastig** [comment]
@nick-arm @theRealAph @stooart-mon 
Hi, could you have a look please?

**Comment by theRealAph** [comment]
> @nick-arm @theRealAph @stooart-mon Hi, could you have a look please?

This is way too complicated. I'd use `MacroAssembler::spin_wait()` to generate a stub and call it from `SpinPause`.

File: src/hotspot/cpu/aarch64/spin_wait_aarch64.hpp [hunk 1]
```diff
@@ -25,24 +25,57 @@
...
+  static void run_yield(int count) {
+    while (count-- > 0) {
+      __asm volatile("yield");
+    }
+  }
+
+  static void run_none(int) {}
 
+  SpinWait(Inst inst = NONE, int count = 0, InstRunner inst_runner = run_none) :
+      _inst(inst), _count(count), _inst_runner(inst_runner)  {}
```
Comment (stooart-mon) [inline src/hotspot/cpu/aarch64/spin_wait_aarch64.hpp:71]: Wouldn't it make more sense to have _inst_runner initialized in the constructor based on the value of Inst inst?
You aren't differentiating between the two in get_spin_wait_desc anyway.

**Comment by eastig** [comment]
> > @nick-arm @theRealAph @stooart-mon Hi, could you have a look please?
> 
> This is way too complicated. I'd use `MacroAssembler::spin_wait()` to generate a stub and call it from `SpinPause`.

Hi @theRealAph,

Thank you for advice. It was a good exercise to learn how to write a stub generator.
I reimplemented SpinPause to use a generated stub.

File: src/hotspot/os_cpu/linux_aarch64/os_linux_aarch64.cpp [hunk 2]
```diff
@@ -382,7 +382,14 @@ int os::extra_bang_size_in_bytes() {
 
 extern "C" {
   int SpinPause() {
-    return 0;
+    if (VM_Version::spin_wait_desc().inst() == SpinWait::NONE) {
+      return 0;
+    }
```
Comment (stooart-mon) [inline src/hotspot/os_cpu/linux_aarch64/os_linux_aarch64.cpp:387]: Would it be safe and more efficient to test func for NULL?

Comment (eastig) [reply src/hotspot/os_cpu/linux_aarch64/os_linux_aarch64.cpp:387]: Good idea.

Comment (eastig) [reply src/hotspot/os_cpu/linux_aarch64/os_linux_aarch64.cpp]: Done

File: src/hotspot/os_cpu/linux_aarch64/os_linux_aarch64.cpp [hunk 3]
```diff
@@ -382,7 +382,13 @@ int os::extra_bang_size_in_bytes() {
 
 extern "C" {
   int SpinPause() {
-    return 0;
+    using spin_wait_func_ptr_t = void (*)();
+    spin_wait_func_ptr_t func = CAST_TO_FN_PTR(spin_wait_func_ptr_t, StubRoutines::aarch64::spin_wait());
+    if (func == nullptr) {
+      return 0;
+    }
```
Comment (theRealAph) [inline src/hotspot/os_cpu/linux_aarch64/os_linux_aarch64.cpp:389]: It's better simply to give `_spin_wait` a default value that points to a `ret` instruction.

File: src/hotspot/os_cpu/linux_aarch64/os_linux_aarch64.cpp [hunk 4]
```diff
@@ -382,7 +382,13 @@ int os::extra_bang_size_in_bytes() {
 
 extern "C" {
   int SpinPause() {
-    return 0;
+    using spin_wait_func_ptr_t = void (*)();
```
Comment (theRealAph) [inline src/hotspot/os_cpu/linux_aarch64/os_linux_aarch64.cpp:385]: I think you'll want `ThreadWXEnable wx(WXExec, thread);` for Apple here.

Comment (eastig) [reply src/hotspot/os_cpu/linux_aarch64/os_linux_aarch64.cpp:385]: Sorry, why do we need it? This is linux-aarch64 implementation of SpinPause.
I think it should in SpinPause in `os_cpu/bsd_aarch64/os_bsd_aarch64.cpp`. Am I right?

File: src/hotspot/os_cpu/linux_aarch64/os_linux_aarch64.cpp (hunk 3 above)
Comment (eastig) [reply src/hotspot/os_cpu/linux_aarch64/os_linux_aarch64.cpp]: Done

File: src/hotspot/os_cpu/linux_aarch64/os_linux_aarch64.cpp (hunk 4 above)
Comment (theRealAph) [reply src/hotspot/os_cpu/linux_aarch64/os_linux_aarch64.cpp:385]: You are indeed right. We can worry about Apple another day.

**Comment by theRealAph** [comment]
Two more changes. The first makes the code simpler, and the second makes it less fragile.

```
diff --git a/src/hotspot/cpu/aarch64/stubGenerator_aarch64.cpp b/src/hotspot/cpu/aarch64/stubGenerator_aarch64.cpp
index a8b2820bb62..e946f3be970 100644
--- a/src/hotspot/cpu/aarch64/stubGenerator_aarch64.cpp
+++ b/src/hotspot/cpu/aarch64/stubGenerator_aarch64.cpp
@@ -6403,9 +6403,7 @@ class StubGenerator: public StubCodeGenerator {
     StubCodeMark mark(this, "StubRoutines", "spin_wait");
     address start = __ pc();
 
-    if (VM_Version::spin_wait_desc().inst() != SpinWait::NONE) {
-      __ spin_wait();
-    }
+    __ spin_wait();
     __ ret(lr);
 
     return start;
diff --git a/src/hotspot/cpu/aarch64/stubRoutines_aarch64.cpp b/src/hotspot/cpu/aarch64/stubRoutines_aarch64.cpp
index bb1a3325cea..f7c27ea7380 100644
--- a/src/hotspot/cpu/aarch64/stubRoutines_aarch64.cpp
+++ b/src/hotspot/cpu/aarch64/stubRoutines_aarch64.cpp
@@ -57,7 +57,10 @@ address StubRoutines::aarch64::_string_indexof_linear_uu = NULL;
 address StubRoutines::aarch64::_string_indexof_linear_ul = NULL;
 address StubRoutines::aarch64::_large_byte_array_inflate = NULL;
 address StubRoutines::aarch64::_method_entry_barrier = NULL;
-address StubRoutines::aarch64::_spin_wait = NULL;
+
+static void spin_wait_nop() { }
+address StubRoutines::aarch64::_spin_wait = CAST_FROM_FN_PTR(address, spin_wait_nop);
+
 bool StubRoutines::aarch64::_completed = false;
 
 /**

```

**Comment by eastig** [comment]
> Two more changes. The first makes the code simpler, and the second makes it less fragile.
> 

Thank you! I forgot `__ spin_wait` generates nothing with `XX:OnSpinWaitInst=none`.

Done.

**Review by theRealAph — APPROVED** [review]

**Review by phohensee — APPROVED** [review]

**Pull request closed without merging**